Project 2 improve calculator

ggcarmi opened the pull request.

ggcarmi committed:
task 3 - add setOperation func that can recive variables

ggcarmi committed:
add evaluate function

there are a lot of changes from the old version of the calculator.

now - all the logic - based on evaluate function

ggcarmi committed:
fix display description problem

it shows withou ... or = sign

ggcarmi committed:
another description fix

ggcarmi committed:
fix storyBoard connect to wrong functions

ggcarmi committed:
fix display bug, add M display label

ggcarmi committed:
clean up, delete comments

ggcarmi committed:
add undo button

ggcarmi committed:
add icon for the app

ggcarmi committed:
clean up

oded-regev commented on MyCalculator/CalculatorBrain.swift:
```diff
@@ -19,38 +19,41 @@ struct CalculatorBrain{
     
     // to store the first operand with the function
-    private var pendingBinaryOperation: PendingBinaryOperation?
+    var pendingBinaryOperation: PendingBinaryOperation?
```
why did you change this from `private`?

ggcarmi replied:
i dont remember if i had a reason, i'll check it

oded-regev commented on MyCalculator/ViewController.swift:
```diff
@@ -58,26 +88,61 @@ class ViewController: UIViewController {
+        }
+        
+        userIsInMiddleOfTyping = false
+        displayValueToUpdate = brain.evaluate(using: variablesDictionary)
```
I'm not sure we should call `evaluate()` here... please verify

ggcarmi replied:
according to the task it should be before, i'll check it

oded-regev commented on MyCalculator/ViewController.swift:
```diff
@@ -58,26 +88,61 @@ class ViewController: UIViewController {
+    
+    @IBAction func getM(_ sender: UIButton) {
+        // M: current value of the display
+        variablesDictionary = ["M": displayValue]
```
Please note that you override the `var variablesDictionary` in this line. If in the future you'll have multiple vars this line will probably cause an unexpected bug since we would expect to **add** the new M entry to the dictionary and not override the entire dictionary

ggcarmi replied:
i know, this is the requirement in the task.
first, i just override the value of M,
but than i notice that the task says to drop the whole dictionary and use a new one.

oded-regev replied:
right.. so in `clear()` I would do something like:
`variablesDictionary = [String: Double]()`

instead of

`variablesDictionary = nil`

And in `getM` just do 
`variablesDictionary['M'] = displayValue`

ggcarmi replied:
agree, i'll change it

oded-regev commented:
I just went over your code and also deeply tested the app. All in all it works great and as expected. The code looks good as well. I put some comments in the code, other than that there are 2 additional issues:

1. Please avoid using ! in the code, use `if let ...` or `guard let ..` 

2. The following scenarios are not working as expected (or at least I think it's not)

- 9, +, undo, 3, = (show 3 instead of 93)
- 3, + , 2, undo, undo, *, 2 (the operator * is ignored)
- 3, +, 2, =, undo, undo (still shows +)
- 2, ->M, 3 + M (displays 2 instead of M)
- 22.11 + 0.5 = , undo, 2, =, undo, undo (shows 0.5 which was deleted)

ggcarmi commented:
1. i will check for the !, it will take some time to get use to it.
2. most of the cases is the same issue and i know what is the problem.
    i think it should act this way, because this is the way he build the calculator.
    in task 10 it says: if the user is in the middle of typing - delete the last character.
    if not - undo the last action.
    so when you going back and you get a number - you are not in the middle of typing.
    i can try to change it (using that boolean) - but im not sure how it will behave.
3.the example with the * is wired, i need to debug this.
4. your last example, it remember the 0.5 because you didnt erased it.
    your first undo cancel the = and not the 0.5

oded-regev commented:
About the "problematic scenarios", can you please let me know for each of the 5 scenarios I've mentioned, which one you think is OK and which one needed fix (and of course provide the fix when needed).

ggcarmi commented:
9, +, undo, 3, = (show 3 instead of 93)
3, + , 2, undo, undo, *, 2 (the operator * is ignored)
3, +, 2, =, undo, undo (still shows +)
2, ->M, 3 + M (displays 2 instead of M)
22.11 + 0.5 = , undo, 2, =, undo, undo (shows 0.5 which was deleted)

1.  the problematic scenario
2. need to check - not sure
3. the problematic scenario - you need one more undo. the first one cancel the =,the second cancel 
   the 2, so its ok to stay with the +. the next undo will do it.
4. it display problem.the calculation are ok.
5. the problematic scenario - you need one more undo. same explanation as the above (3)

ggcarmi committed:
bug fixes

change all ! to guard/if-let
change vars to private
dictionary re-init in clear()

ggcarmi committed:
fix bug - undo when the screen is empty

if the screen is empty, cancel the last operation
